### PR TITLE
Add xcode_server ci_source;

### PR DESCRIPTION
**Pull request by Antondomashnev**
Feature request from issue #171 
This will only work if use [Buildasaur](https://github.com/czechboy0/Buildasaur) to build PR on Xcode Server.

### CODE REVIEW

**Comment by KrauseFx** [comment]
This is missing some docs

File: lib/danger/ci_source/xcode_server.rb [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+module Danger
+  module CISource
+    class XcodeServer < CI
+      def self.validates?(env)
+        return !env["XCS_BOT_NAME"].nil?
+      end
+
+      def initialize(env)
+        bot_name = env["XCS_BOT_NAME"]
+        unless bot_name.nil?
```
Comment (jeroenvisser101) [inline lib/danger/ci_source/xcode_server.rb]: you might want to do `return if bot_name.nil?` instead here

Comment (Antondomashnev) [inline lib/danger/ci_source/xcode_server.rb]: @jeroenvisser101 thanks, that is make sense.

File: Dangerfile [hunk 2]
```diff
@@ -1,40 +1,13 @@
...
-if modified_files.empty? && added_files.empty? && deleted_files.empty?
-  fail "This PR has no changes at all, this is likely a developer issue."
-end
-
-protect_files(path: "danger.gemspec",
-           message: ".gemspec modified",
-        fail_build: false)
+# Don't let testing shortcuts get into master by accident
+fail("fdescribe left in tests") if `grep -r fdescribe specs/`.length > 1
+fail("fit left in tests") if `grep -r "fit specs/ `.length > 1
```
Comment (orta) [inline Dangerfile]: any chance you could undo this change?

File: lib/danger/ci_source/xcode_server.rb [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+
+      def initialize(env)
+        bot_name = env["XCS_BOT_NAME"]
+        return if bot_name.nil?
+
+        repo_matches = bot_name.match(/\[(.*?)\]/)
+        self.repo_slug = repo_matches[1] unless repo_matches.nil?
+        pull_request_id_matches = bot_name.match(/((?<=#)\d+)/)
+        self.pull_request_id = pull_request_id_matches[1] unless pull_request_id_matches.nil?
+      end
```
Comment (orta) [inline lib/danger/ci_source/xcode_server.rb]: :+1:

**Comment by orta** [comment]
Looks good

File: Dangerfile (hunk 2 above)
Comment (jeroenvisser101) [inline Dangerfile]: Possibly we could add this version to a `example/xcode/Dangerfile` so it would serve as quick-start, and we'd do the same with other providers/integrations? But this change should be undone indeed.

Comment (Antondomashnev) [inline Dangerfile]: Sure, my mistake sorry.

Comment (orta) [inline Dangerfile]: This was @Antondomashnev very reasonably running `bin/danger init` to test - we could look in the command to see if you're actually in the Danger git repo and not make changes to the existing `Dangerfile`?

File: lib/danger/ci_source/xcode_server.rb [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+    class XcodeServer < CI
+      def self.validates?(env)
+        return !env["XCS_BOT_NAME"].nil?
+      end
+
+      def initialize(env)
+        bot_name = env["XCS_BOT_NAME"]
+        return if bot_name.nil?
+
+        repo_matches = bot_name.match(/\[(.*?)\]/)
```
Comment (jeroenvisser101) [inline lib/danger/ci_source/xcode_server.rb]: this regex should probably be `\[(.+)\]`, since it may never be empty.

File: lib/danger/ci_source/xcode_server.rb [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+        self.repo_slug = repo_matches[1] unless repo_matches.nil?
+        pull_request_id_matches = bot_name.match(/((?<=#)\d+)/)
```
Comment (jeroenvisser101) [inline lib/danger/ci_source/xcode_server.rb]: Do you need a positive lookbehind here? Wouldn't `/#(\d+)/` not work on it's own?

File: Dangerfile (hunk 2 above)
Comment (jeroenvisser101) [inline Dangerfile]: Or change our Dangerfile to warn when the `Dangerfile` is changed, and explain when this is allowed, and when it's not. Either way, we could have some safeguards there for sure!

Comment (orta) [inline Dangerfile]: Our Dangerfile would be overwritten, and that's what'd get parsed - so it wouldn't be able to tell itself to fail :dancers: 

But I agree on the concept :+1:

Comment (jeroenvisser101) [inline Dangerfile]: Shit, you're right haha, that'd be bad. Yeah, we could probably change `danger init` and check if `danger/danger` is in the output of `git remote -v` (or something similar)?

File: lib/danger/ci_source/xcode_server.rb (hunk 4 above)
Comment (Antondomashnev) [inline lib/danger/ci_source/xcode_server.rb]: yes, you're right

File: lib/danger/ci_source/xcode_server.rb (hunk 5 above)
Comment (Antondomashnev) [inline lib/danger/ci_source/xcode_server.rb]: not really, if the [Buildasaur](https://github.com/czechboy0/Buildasaur) pattern has only one `#`. I think we can omit positive lookbehind here

Comment (jeroenvisser101) [inline lib/danger/ci_source/xcode_server.rb]: 👍

**Comment by jeroenvisser101** [comment]
Looks good to me!

**Comment by orta** [comment]
Yep - same here, great job @Antondomashnev

**Comment by czechboy0** [comment]
Great work! Just one piece of feedback:
Currently when someone uses Xcode Server but doesn't use Buildasaur, the validates function seems to return true in that case. I suggest you try to parse the bot name, and if it's there, also check that it parses into the two expected parts (a string and a number). Only then it should validate. 

You could even print a warning if someone does try to run danger on Xcode Server without Buildasaur, maybe even just a link to this issue would be helpful.
